### PR TITLE
[DNM] Initial release of nrf_oberon and nrf_security documentation

### PR DESCRIPTION
**Pull request by frkv**
- Added API documentation for nrf_oberon
    - Added API documentation for nrf_security
    - Added License headers in c-files/h-files
    - RSA: Renamed export to export_fn as it is a reserved keyword in C++

    Signed-off-by: Frank Audun Kvamtrø <frank.kvamtro@nordicsemi.no>

### CODE REVIEW

File: crypto/nrf_oberon/include/ocrypto_constant_time.h [hunk 1]
```diff
@@ -5,6 +5,17 @@
  */
 
 /**@file
+ * @defgroup nrf_oberon Oberon cryptographic library
+ * @{
+ * @brief Highly optimized cryptographic algorithms implementation for Cortex-M0, Cortex-M4
```
Comment (oyvindronningstad) [inline crypto/nrf_oberon/include/ocrypto_constant_time.h]: algorithms -> algorithm
optionally: implementation -> implementations

File: crypto/nrf_oberon/include/ocrypto_curve_p256.h [hunk 2]
```diff
@@ -15,38 +23,75 @@ extern "C" {
...
     ocrypto_mod_p256 y;
 } ocrypto_cp_p256;
+/**@endcond */
 
-
-// load r.x from bytes, recover r.y
-// returns 0 if r is a legal curve point, -1 otherwise
+/** load r.x from bytes, recover r.y
+ *
+ * @param[out]      r       Point with r.x loaded, r.y kept
```
Comment (oyvindronningstad) [inline crypto/nrf_oberon/include/ocrypto_curve_p256.h]: This seems to contradict the brief. Is r.y recovered or kept by this function?

File: crypto/nrf_oberon/include/ocrypto_curve_p256.h [hunk 3]
```diff
@@ -15,38 +23,75 @@ extern "C" {
...
+ * @param           p       Point as array of bytes
+ *
+ * @returns 0 if r is a legal curve point, -1 otherwise
+ */
 int ocrypto_curve_p256_from64bytes(ocrypto_cp_p256 *r, const uint8_t p[64]);
 
-// store p.x to bytes
+/** store p.x to bytes
+ *
+ * @param[out]           r       x stored as array
```
Comment (oyvindronningstad) [inline crypto/nrf_oberon/include/ocrypto_curve_p256.h]: x or p.x?

File: crypto/nrf_oberon/include/ocrypto_curve_p256.h [hunk 4]
```diff
@@ -15,38 +23,75 @@ extern "C" {
...
+ * @param                p       Point with x to be stored
+ */
 void ocrypto_curve_p256_to32bytes(uint8_t r[32], ocrypto_cp_p256 *p);
 
-// store point to bytes
+/** store p.x to bytes
+ *
+ * @param[out]           r       Pont stored as array
```
Comment (oyvindronningstad) [inline crypto/nrf_oberon/include/ocrypto_curve_p256.h]: Pont -> Point

File: crypto/nrf_oberon/include/ocrypto_curve_p256.h [hunk 5]
```diff
@@ -15,38 +23,75 @@ extern "C" {
...
+ */
 void ocrypto_curve_p256_to64bytes(uint8_t r[64], ocrypto_cp_p256 *p);
 
-// r = p * s
-// r = [0,0] if p = [0,0] or s mod q = 0
-// returns -1 if r = [0,0], 0 if 0 < s < q, 1 if s > q
+/** P256 scalar multiplication
+ *
+ * r = p * s
+ * r = [0,0] if p = [0,0] or s mod q = 0
```
Comment (oyvindronningstad) [inline crypto/nrf_oberon/include/ocrypto_curve_p256.h]: Where does q come from?

File: crypto/nrf_oberon/include/ocrypto_ecdh_p256.h [hunk 6]
```diff
@@ -14,19 +21,25 @@ extern "C" {
 #include <stdint.h>
 
 
-// ecdh P-256 public key r = n * p
-// r[64]: the resulting public key
-// s[32]: the secret key
-// r may be same as s
-// returns 0 if s is a legal secret key
+/** ecdh P-256 public key r = n * p
```
Comment (oyvindronningstad) [inline crypto/nrf_oberon/include/ocrypto_ecdh_p256.h]: Calculate ecdh P-256 public key from private key

File: crypto/nrf_oberon/include/ocrypto_ecdh_p256.h [hunk 7]
```diff
@@ -14,19 +21,25 @@ extern "C" {
...
+ */
 int ocrypto_ecdh_p256_public_key(uint8_t r[64], const uint8_t s[32]);
 
-// ecdh P-256 common secret
-// r[32]: the resulting secret
-// s[32]: the secret key
-// p[64]: the public key
-// r may be same as s or p
-// returns 0 if s is a legal secret key and p is a legal public key
+/** ecdh P-256 common secret
```
Comment (oyvindronningstad) [inline crypto/nrf_oberon/include/ocrypto_ecdh_p256.h]: Is this the same as shared secret? If so say "ecdh P-256 common (shared) secret"

File: crypto/nrf_oberon/include/ocrypto_ecdsa_p256.h [hunk 8]
```diff
@@ -13,55 +20,72 @@ extern "C" {
...
 #include <stdint.h>
 
-
-// ecdsa P-256 signature key pair generation
-// pk[64]: the generated public key
-// sk[32]: the private key
-// returns 0 if sk is a legal secret key
+/**@brief ecdsa P-256 signature key pair generation
+ * @param[out]  pk  The generated public key
+ * @param       sk  The Private key
```
Comment (oyvindronningstad) [inline crypto/nrf_oberon/include/ocrypto_ecdsa_p256.h]: Is this param in or out? The brief suggests it's out

File: crypto/nrf_oberon/include/ocrypto_ecdsa_p256.h [hunk 9]
```diff
@@ -13,55 +20,72 @@ extern "C" {
...
     const uint8_t sk[32]);
 
-// ecdsa P-256 signature generate
-// sig[64]: the generated signature
-// m[mlen]: the input message
-// sk[32]:  the secret key
-// ek[32]:  the session key
-// returns 0 if ek is a valid session key
+
+/**@brief ecdsa P-256 signature generate
```
Comment (oyvindronningstad) [inline crypto/nrf_oberon/include/ocrypto_ecdsa_p256.h]: "Generate ecdsa P-256 signature"

File: crypto/nrf_oberon/include/ocrypto_ecdsa_p256.h [hunk 10]
```diff
@@ -13,55 +20,72 @@ extern "C" {
 #include <stddef.h>
 #include <stdint.h>
 
-
-// ecdsa P-256 signature key pair generation
-// pk[64]: the generated public key
-// sk[32]: the private key
-// returns 0 if sk is a legal secret key
+/**@brief ecdsa P-256 signature key pair generation
```
Comment (oyvindronningstad) [inline crypto/nrf_oberon/include/ocrypto_ecdsa_p256.h]: Earlier files omit "`@brief`"

File: crypto/nrf_oberon/include/ocrypto_ecdsa_p256.h [hunk 11]
```diff
@@ -13,55 +20,72 @@ extern "C" {
...
+
+/**@brief ecdsa P-256 signature generate
+ *
+ * @param[out]  sig     The generated signature
+ * @param       m       The input message
+ * @param       mlen    Length of the message
+ * @param       sk      The secret key
+ * @param       ek      The session key
+ *
+ * @returns 0 if ek is a valid session key
```
Comment (oyvindronningstad) [inline crypto/nrf_oberon/include/ocrypto_ecdsa_p256.h]: Is `@returns` valid doxygen?

Comment (oyvindronningstad) [reply crypto/nrf_oberon/include/ocrypto_ecdsa_p256.h]: This and other return value dox in this file seem to omit a lot of info.

File: crypto/nrf_oberon/include/ocrypto_ecdsa_p256.h [hunk 12]
```diff
@@ -13,55 +20,72 @@ extern "C" {
...
     const uint8_t ek[32]);
 
-// ecdsa P-256 signature generate
-// sig[64]: the generated signature
-// hash[32]: SHA-256 hash of the input message
-// sk[32]:  the secret key
-// ek[32]:  the session key
-// returns 0 if ek is a valid session key
+
+/**@brief ecdsa P-256 signature generate from hash
```
Comment (oyvindronningstad) [inline crypto/nrf_oberon/include/ocrypto_ecdsa_p256.h]: generate ecdsa P-256 signature from hash

File: crypto/nrf_oberon/include/ocrypto_constant_time.h (hunk 1 above)
Comment (frkv) [reply crypto/nrf_oberon/include/ocrypto_constant_time.h]: Done

File: crypto/nrf_oberon/include/ocrypto_curve_p256.h (hunk 2 above)
Comment (frkv) [reply crypto/nrf_oberon/include/ocrypto_curve_p256.h]: Made this clearer

File: crypto/nrf_oberon/include/ocrypto_curve_p256.h (hunk 5 above)
Comment (frkv) [reply crypto/nrf_oberon/include/ocrypto_curve_p256.h]: From the curve

File: crypto/nrf_oberon/include/ocrypto_ecdsa_p256.h (hunk 8 above)
Comment (frkv) [reply crypto/nrf_oberon/include/ocrypto_ecdsa_p256.h]: It should be out as well, yes

File: crypto/nrf_oberon/include/ocrypto_ecdsa_p256.h (hunk 10 above)
Comment (frkv) [reply crypto/nrf_oberon/include/ocrypto_ecdsa_p256.h]: Will remove for consistency

File: crypto/nrf_oberon/include/ocrypto_ecdsa_p256.h (hunk 11 above)
Comment (frkv) [reply crypto/nrf_oberon/include/ocrypto_ecdsa_p256.h]: \returns is valid doxygen, rest will be reported to content creators

File: crypto/nrf_oberon/include/ocrypto_ecdsa_p256.h (hunk 12 above)
Comment (frkv) [reply crypto/nrf_oberon/include/ocrypto_ecdsa_p256.h]: Made a small alteration

**Comment by frkv** [comment]
Closing this PR and opening this in an additional PR (due to squashed rebase commits)

**Pull request closed without merging**